quantization/src/encoded_vectors_pq.rs

```diff
@@ -183,6 +207,62 @@ impl<TStorage: EncodedStorage> EncodedVectorsPQ<TStorage> {
         }
     }
 
+    #[allow(unused)]
+    fn encode_storage_rayon<'a, SB>(
+        data: impl ParallelIterator<Item = &'a [f32]>,
+        storage_builder: Arc<Mutex<&mut SB>>,
+        _vector_parameters: &VectorParameters,
+        vector_division: &[Range<usize>],
+        centroids: &[Vec<f32>],
+        max_kmeans_threads: usize,
+    ) where
+        SB: EncodedStorageBuilder<TStorage> + Send,
+    {
+        let pool = rayon::ThreadPoolBuilder::new()
+            .thread_name(|idx| format!("pq-{idx}"))
+            .num_threads(max_kmeans_threads)
+            .build()
+            .map_err(|e| EncodingError {
+                description: format!("Failed PQ encoding while thread pool init: {e}"),
+            })
+            .unwrap();
+
+        // TODO: minimize allocation, use buffer types
```

Comment:
The trick you used to minimize allocation by reusing `Vec`s can be added again here. But I didn't take the time to convert it to rayon.